src/passes/orderNestedStructs.ts

```diff
@@ -0,0 +1,123 @@
+import {
+  ArrayType,
+  ContractDefinition,
+  getNodeType,
+  MappingType,
+  SourceUnit,
+  StructDefinition,
+  TypeNode,
+  UserDefinedType,
+} from 'solc-typed-ast';
+import { AST } from '../ast/ast';
+import { ASTMapper } from '../ast/mapper';
+
+export class OrderNestedStructs extends ASTMapper {
+  // Cairo does not permit to use struct definitions which are yet to be defined.
+  // For example:
+  // contract Warp {
+  //    struct Top {
+  //        Nested n;
+  //    }
+  //    struct Nested { ... }
+  // }
+  // When transpiled to Cairo, struct definitions must be reordered so that
+  // nested structs are defined first:
+  //   struct Nested { ... }
+  //   struct Top {
+  //     member n : Nested;
+  //   }
+
+  visitSourceUnit(node: SourceUnit, ast: AST): void {
+    this.reorderNestedStructs(node, ast);
+    this.commonVisit(node, ast);
+  }
+
+  visitContractDefinition(node: ContractDefinition, ast: AST): void {
+    this.reorderNestedStructs(node, ast);
+  }
+
+  private reorderNestedStructs(node: SourceUnit | ContractDefinition, ast: AST) {
+    const structs = node.vStructs;
+    const [roots, tree] = this.makeStructTree(structs, ast);
+
+    // there are no nested structs
+    if (roots.size === structs.length) return;
+
+    const newStructOrder = this.reorderStructs(roots, tree);
+
+    // remove old struct definiton
+    structs.forEach((child) => {
+      if (child instanceof StructDefinition) {
+        node.removeChild(child);
+      }
+    });
+
+    // insert back in new order
+    newStructOrder.reverse().forEach((struct) => node.insertAtBeginning(struct));
+  }
+
+  private makeStructTree(
+    structs: readonly StructDefinition[],
+    ast: AST,
+  ): [Set<StructDefinition>, Map<StructDefinition, [StructDefinition]>] {
+    const roots = new Set<StructDefinition>(structs);
+    const tree = new Map<StructDefinition, [StructDefinition]>();
```

Comment:
Assuming you want this to be a one to many struct definition mapping it should be `Map<StructDefinition, StructDefinition[]>`, rather than the current map to arrays of length 1